### PR TITLE
Separate company data from code in company model

### PR DESCRIPTION
**Pull request by cykhoo**
This has been bugging me for some time now, and I've only found the time to do some refactoring today ;).

This pull request extracts the company data previously stored in the LIST constant into a separate yaml file (companies.yml), which is read in and exposed in a new companies method.

This separates the data of the companies from the code in the company model, which should deal with the business concerns of the company model. It also allows the data to be updated more easily.

It's my first proper pull request, so please be gentle. All feedback gratefully received!

### CODE REVIEW

File: app/models/company.rb [hunk 1]
```diff
@@ -18,299 +20,21 @@ class Company
...
-      }
-    ]
+      return @companies if @companies
+
+      File.open(File.join(File.dirname(__FILE__), "companies.yml")) do |file|
+        @companies = YAML.load(file).to_ary
+      end
+
+      @companies
+    end
```
Comment (JuanitoFatas) [inline app/models/company.rb]: How about this style? :blush: 

```ruby
def companies
  @_companies ||= begin
    File.open(File.join(File.dirname(__FILE__), "companies.yml")) do |file|
      @companies = YAML.load(file).to_ary
    end
  end
end
```

Comment (cykhoo) [inline app/models/company.rb]: Yes, of course. I knew something wasn't quite as elegant as it should be. Will make the fix.

**Comment by JuanitoFatas** [comment]
Personally I prefer to use `self` for class methods. :smile:

**Comment by Drenmi** [comment]
Great! I am all for this change, in principle. :smile: Here's some feedback on your first PR:

1. The `README` with instructions for how to add your company would need to be updated to reflect the new location of the data.

2. The YAML file will be read from the filesystem and parsed on each request. Normally this is done in an initializer. Also `.companies` does not need to be part of the public interface of the class.

3. When contributing to an existing codebase, it is best to assume that any conventions have been put in place deliberately by the maintainer, and follow them, even if it is not what we'd normally do. (In this case the syntax for defining class methods.)

:kissing_heart:

**Comment by JuanitoFatas** [comment]
> The YAML file will be read from the filesystem and parsed on each request. Normally this is done in an initializer. Also .companies does not need to be part of the public interface of the class.

:+1: on this too!

Some hints :wink: 

```ruby
# config/initializers/constants.rb
module RubysgReboot
  COMPANIES_LIST = # load your yaml...
end

# Use in codebase
RubysgReboot::COMPANIES_LIST

# In tests, you can stub the constant like this before the spec run
stub_const("RubysgReboot::COMPANIES_LIST", "load the yaml here")
```

**Comment by cykhoo** [comment]
@Drenmi and @JuanitoFatas thanks for the feedback. Will do some work over the weekend. I was hesitant to make the companies method public, but that was the only way I could have the method stubbing in the example work. How would you do this otherwise?

**Comment by Drenmi** [comment]
@cykhoo: The guideline is that objects should tell each other what they want (e.g. "give me all companies"), but not how the receiver does it. Outside objects shouldn't know about the method loading the file.

How to test is a good question, but maybe before we can ask what we should test. I do more behaviour style testing, so I would probably test that the output from `Company.all` is what I expect, and ignore the implementation details.

By testing the internal method calls, we can't refactor the class without also changing the tests, which is normally quite dangerous. (You generally want to change only one of them at a time.)

**Comment by cykhoo** [comment]
Khello. Thanks @JuanitoFatas and @Drenmi for the great tips!

File: app/models/company.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require 'yaml'
```
Comment (winston) [inline app/models/company.rb]: Might be more accurate to require this in the initializer instead.

File: config/initializers/constants.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module RubysgReboot
+
```
Comment (winston) [inline config/initializers/constants.rb]: :scissors: empty line.

**Comment by winston** [comment]
@cykhoo Awesome! Thanks for this!

Just one more comment for discussion:
It's ok to have the list loaded as one of the initializers. I might consider to put the `.yml` file somewhere else though.  @Drenmi any suggestions on where you would usually put this? Maybe the config folder?

Thanks!

File: app/models/company.rb (hunk 2 above)
Comment (cykhoo) [inline app/models/company.rb]: Yup!

**Comment by cykhoo** [comment]
@winston I wonder if we are approaching this the wrong way, holding on to the list of companies in a constant rather than reading it from the database.

If we take the latter approach, then perhaps a rake task in /lib/tasks/ run on deploy to load from the yaml file and write to the database might be the way to go? The yaml file would then live in /lib/tasks/ too.

**Comment by winston** [comment]
> @winston I wonder if we are approaching this the wrong way, holding on to the list of companies in a constant rather than reading it from the database.

@cykhoo I am really open to any kind of change, so long as people can still add a company via a PR (because I would like people to participate through geeky means vs a form haha), and easy deployment. 

Currently this app is deployed to Heroku automatically by Heroku once master passes the specs, and unfortunately, I don't think there's a way to inject the additional rake task though (the automation also doesn't run db:migrate, so if there's a db change, I'll have to do it manually after deployment).

**Comment by winston** [comment]
Anyway, this is good to merge for me. @Drenmi @JuanitoFatas ?

**Comment by Drenmi** [comment]
Sorry for not responding earlier. I really need to configure my notifications in GitHub ...

Where would I put it? Hm. Good question. Most likely in `config/locales`, since this is where other textual content is stored.

Regarding the PR itself, I have no objections. The only thing I would add is freezing the constant, but a bit late for that now. :smile:

**Comment by winston** [comment]
> The only thing I would add is freezing the constant, but a bit late for that now.

@Drenmi Another PR then if it's really bothering you. :)

**Comment by winston** [comment]
> Most likely in config/locales

Where @cykhoo has placed it now works for me too, so don't think this is necessary for now.

**Comment by arnvald** [comment]
Since this YAML file is kind of a database (and yeah, YAML is used for data storage: https://github.com/nicotaing/yaml_record) it probably should be kept in `/db` directory. The same happens when you use SQLite as your db, you simply keep the file in `/db`.

However, as @winston said, as long as it works, it's ok :+1:

**Pull request merged**